### PR TITLE
Register brinkman_time_independent flag for each Q variable

### PR DESCRIPTION
**Pull request by Raamkrishnan**
### IBAMR Pull Request Checklist
- [x] Does the test suite pass?
- [x] Was clang-format run?
- [x] Were relevant issues cited? Please remember to add `Fixes #12345` to close
      the issue automatically if we are fixing the problem.
- [x] Is this a change others will want to know about? If so, then has a
      changelog entry been added?
- [x] If new data structures have been added to a class then have they been set
      up appropriately for restarts? If so, ensure that the restart version
      number is incremented.
- [x] Does this change include a bug fix or new functionality? If so, a new test
      or tests should be added. New tests should run quickly (less than a minute
      in release mode). If possible, an older test should gain a new option so
      that we do not need to compile more test executables.
- [x] Did you (if your account has permission to do so) set relevant labels on
      GitHub for the pull request?

@amneetb As we discussed, now I set the brinkman_time_independent flag for each Q variable. The default is set to false. With this, the current implementation can solve both Brinkman time-dependent and time-independent problems.

### CODE REVIEW

File: include/ibamr/BrinkmanAdvDiffSemiImplicitHierarchyIntegrator.h [hunk 1]
```diff
@@ -150,6 +150,23 @@ class BrinkmanAdvDiffSemiImplicitHierarchyIntegrator : public AdvDiffSemiImplici
         return d_brinkman_penalization;
     } // getBrinkmanPenalization
 
+    /*!
+     * \brief Set the boolean to be TRUE for the time independent Brinkman problems.
+     * Default option is false.
+     */
+    void setBrinkmanTimeIndepedent(SAMRAI::tbox::Pointer<SAMRAI::pdat::CellVariable<NDIM, double> > Q_var,
```
Comment (amneetb) [inline include/ibamr/BrinkmanAdvDiffSemiImplicitHierarchyIntegrator.h:157]: suggest renaming it to ` setBrinkmanVariableTimeIndepedent`

File: include/ibamr/BrinkmanAdvDiffSemiImplicitHierarchyIntegrator.h [hunk 2]
```diff
@@ -150,6 +150,23 @@ class BrinkmanAdvDiffSemiImplicitHierarchyIntegrator : public AdvDiffSemiImplici
         return d_brinkman_penalization;
     } // getBrinkmanPenalization
 
+    /*!
+     * \brief Set the boolean to be TRUE for the time independent Brinkman problems.
+     * Default option is false.
```
Comment (amneetb) [inline include/ibamr/BrinkmanAdvDiffSemiImplicitHierarchyIntegrator.h:155]: \brief Indicate if are solving a time independent problem for the Brinkman variable Q.
     * Default option is false, which means Q is assumed to vary with time.

File: examples/adv_diff/ex7/example.cpp [hunk 3]
```diff
@@ -317,6 +317,8 @@ main(int argc, char* argv[])
         Pointer<BrinkmanAdvDiffSemiImplicitHierarchyIntegrator> bp_adv_diff_hier_integrator = time_integrator;
         bp_adv_diff_hier_integrator->registerBrinkmanAdvDiffBcHelper(brinkman_adv_diff);
 
+        bp_adv_diff_hier_integrator->setBrinkmanTimeIndepedent(q_var, true /* brinkman_time_independent */);
+
```
Comment (amneetb) [inline examples/adv_diff/ex7/example.cpp:321]: bp_adv_diff_hier_integrator->setBrinkmanVariableTimeIndepedent(q_var, /*time_independent*/ true);

File: examples/adv_diff/ex6/example.cpp [hunk 4]
```diff
@@ -519,6 +519,8 @@ main(int argc, char* argv[])
         Pointer<BrinkmanAdvDiffSemiImplicitHierarchyIntegrator> bp_adv_diff_hier_integrator = time_integrator;
         bp_adv_diff_hier_integrator->registerBrinkmanAdvDiffBcHelper(brinkman_adv_diff);
 
+        bp_adv_diff_hier_integrator->setBrinkmanTimeIndepedent(q_var, true /* brinkman_time_independent */);
+
```
Comment (amneetb) [inline examples/adv_diff/ex6/example.cpp:523]: bp_adv_diff_hier_integrator->setBrinkmanVariableTimeIndepedent(q_var, /*time_independent*/ true);

File: examples/adv_diff/ex5/example.cpp [hunk 5]
```diff
@@ -256,6 +256,8 @@ main(int argc, char* argv[])
         Pointer<BrinkmanAdvDiffSemiImplicitHierarchyIntegrator> bp_adv_diff_hier_integrator = time_integrator;
         bp_adv_diff_hier_integrator->registerBrinkmanAdvDiffBcHelper(brinkman_adv_diff);
 
+        bp_adv_diff_hier_integrator->setBrinkmanTimeIndepedent(q_var, true /* brinkman_time_independent */);
+
```
Comment (amneetb) [inline examples/adv_diff/ex5/example.cpp:260]: bp_adv_diff_hier_integrator->setBrinkmanVariableTimeIndepedent(q_var, /*time_independent*/ true);

File: examples/adv_diff/ex4/example.cpp [hunk 6]
```diff
@@ -257,6 +257,8 @@ main(int argc, char* argv[])
         Pointer<BrinkmanAdvDiffSemiImplicitHierarchyIntegrator> bp_adv_diff_hier_integrator = time_integrator;
         bp_adv_diff_hier_integrator->registerBrinkmanAdvDiffBcHelper(brinkman_adv_diff);
 
+        bp_adv_diff_hier_integrator->setBrinkmanTimeIndepedent(q_var, true /* brinkman_time_independent */);
+
```
Comment (amneetb) [inline examples/adv_diff/ex4/example.cpp:261]: bp_adv_diff_hier_integrator->setBrinkmanVariableTimeIndepedent(q_var, /*time_independent*/ true);

File: examples/adv_diff/ex3/example.cpp [hunk 7]
```diff
@@ -217,6 +217,8 @@ main(int argc, char* argv[])
         Pointer<BrinkmanAdvDiffSemiImplicitHierarchyIntegrator> bp_adv_diff_hier_integrator = time_integrator;
         bp_adv_diff_hier_integrator->registerBrinkmanAdvDiffBcHelper(brinkman_adv_diff);
 
+        bp_adv_diff_hier_integrator->setBrinkmanTimeIndepedent(q_var, true /* brinkman_time_independent */);
+
```
Comment (amneetb) [inline examples/adv_diff/ex3/example.cpp:221]: bp_adv_diff_hier_integrator->setBrinkmanVariableTimeIndepedent(q_var, /*time_independent*/ true);

File: src/adv_diff/BrinkmanAdvDiffSemiImplicitHierarchyIntegrator.cpp [hunk 8]
```diff
@@ -856,6 +868,28 @@ BrinkmanAdvDiffSemiImplicitHierarchyIntegrator::registerBrinkmanAdvDiffBcHelper(
     return;
 } // registerBrinkmanAdvDiffBcHelper
 
+void
+BrinkmanAdvDiffSemiImplicitHierarchyIntegrator::setBrinkmanTimeIndepedent(Pointer<CellVariable<NDIM, double> > Q_var,
+                                                                          bool brinkman_time_independent)
```
Comment (amneetb) [inline src/adv_diff/BrinkmanAdvDiffSemiImplicitHierarchyIntegrator.cpp:873]: setBrinkmanVariableTimeIndepedent(Pointer<CellVariable<NDIM, double> > Q_var,
                                                                          bool time_independent);

File: src/adv_diff/BrinkmanAdvDiffSemiImplicitHierarchyIntegrator.cpp [hunk 9]
```diff
@@ -856,6 +868,28 @@ BrinkmanAdvDiffSemiImplicitHierarchyIntegrator::registerBrinkmanAdvDiffBcHelper(
...
+BrinkmanAdvDiffSemiImplicitHierarchyIntegrator::registerTransportedQuantity(Pointer<CellVariable<NDIM, double> > Q_var,
+                                                                            bool output_Q)
+{
+#if !defined(NDEBUG)
+    TBOX_ASSERT(Q_var);
+    TBOX_ASSERT(std::find(d_Q_var.begin(), d_Q_var.end(), Q_var) == d_Q_var.end());
+#endif
+    AdvDiffSemiImplicitHierarchyIntegrator::registerTransportedQuantity(Q_var, output_Q);
+    d_brinkman_time_independent[Q_var] = false;
+    return;
```
Comment (amneetb) [inline src/adv_diff/BrinkmanAdvDiffSemiImplicitHierarchyIntegrator.cpp:889]: Let's use `d_Q_time_independent` instead of  `d_brinkman_time_independent` as it is used for all Q vars.

File: include/ibamr/BrinkmanAdvDiffSemiImplicitHierarchyIntegrator.h [hunk 10]
```diff
@@ -165,7 +182,7 @@ class BrinkmanAdvDiffSemiImplicitHierarchyIntegrator : public AdvDiffSemiImplici
      * Flag to zero out the temporal term contribution when the Brinkman approach
      * is used.
      */
-    bool d_brinkman_time_independent = false;
+    std::map<SAMRAI::tbox::Pointer<SAMRAI::pdat::CellVariable<NDIM, double> >, bool> d_brinkman_time_independent;
```
Comment (amneetb) [inline include/ibamr/BrinkmanAdvDiffSemiImplicitHierarchyIntegrator.h:186]: d_Q_time_independent

File: src/adv_diff/BrinkmanAdvDiffSemiImplicitHierarchyIntegrator.cpp [hunk 11]
```diff
@@ -535,7 +547,7 @@ BrinkmanAdvDiffSemiImplicitHierarchyIntegrator::integrateHierarchy(const double
             // set 1.0/dt - (1.0-K)*Cb in rhs_op_spec, which is applied to Q later
             d_hier_cc_data_ops->scale(Cb_rhs_scratch_idx, -(1.0 - K), Cb_current_idx);
             // In the case of Poisson applications zero out the temporal contribution.
-            if (d_brinkman_time_independent)
+            if (d_brinkman_time_independent[Q_var])
             {
```
Comment (amneetb) [inline src/adv_diff/BrinkmanAdvDiffSemiImplicitHierarchyIntegrator.cpp:551]: use d_Q_time_independent instead.

**Review by amneetb — CHANGES_REQUESTED** [review]

File: src/adv_diff/BrinkmanAdvDiffSemiImplicitHierarchyIntegrator.cpp (hunk 11 above)
Comment (amneetb) [reply src/adv_diff/BrinkmanAdvDiffSemiImplicitHierarchyIntegrator.cpp:551]: BTW, can we do this check for all Q_vars here instead of just those marked as Brinkman one? That would be more consistent.

Comment (amneetb) [reply src/adv_diff/BrinkmanAdvDiffSemiImplicitHierarchyIntegrator.cpp:551]: And in that case `setBrinkmanVariableTimeIndependent()` should be named `setTransportQuantityTimeIndependent()`.

**Comment by Raamkrishnan** [comment]
@amneetb I have made changes in the code based on your suggestions. Now I initialize the solver and RHS operator every time step when the dt or diffusion coefficient changes.

I have also made this change in ```AdvDiffSemiImplicitHierarchyIntegrator``` class.

**Comment by amneetb** [comment]
This looks good to me. Let's wait for @drwells' and @boyceg's review.

**Comment by Raamkrishnan** [comment]
@drwells If you are happy with the PR, can we merge this?

File: include/ibamr/BrinkmanAdvDiffSemiImplicitHierarchyIntegrator.h (hunk 2 above)
Comment (drwells) [reply include/ibamr/BrinkmanAdvDiffSemiImplicitHierarchyIntegrator.h:155]: We still need to make this correction.

File: examples/adv_diff/ex3/example.cpp (hunk 7 above)
Comment (drwells) [reply examples/adv_diff/ex3/example.cpp:221]: We still need to make this correction (etc).

File: src/adv_diff/AdvDiffSemiImplicitHierarchyIntegrator.cpp [hunk 12]
```diff
@@ -574,6 +582,8 @@ AdvDiffSemiImplicitHierarchyIntegrator::preprocessIntegrateHierarchy(const doubl
         helmholtz_rhs_op->setHomogeneousBc(false);
         helmholtz_rhs_op->setSolutionTime(current_time);
         helmholtz_rhs_op->setTimeInterval(current_time, new_time);
+
+        // Reinitializing the RHS operator every time step when the time step size or diffusion coefficient changes.
```
Comment (drwells) [inline src/adv_diff/AdvDiffSemiImplicitHierarchyIntegrator.cpp:586]:
```suggestion
```

File: src/adv_diff/AdvDiffSemiImplicitHierarchyIntegrator.cpp [hunk 13]
```diff
@@ -751,6 +764,8 @@ AdvDiffSemiImplicitHierarchyIntegrator::integrateHierarchy(const double current_
         helmholtz_solver->setHomogeneousBc(false);
         helmholtz_solver->setSolutionTime(new_time);
         helmholtz_solver->setTimeInterval(current_time, new_time);
+
+        // Reinitializing the solver every time step when the time step size or diffusion coefficient changes.
```
Comment (drwells) [inline src/adv_diff/AdvDiffSemiImplicitHierarchyIntegrator.cpp:768]:
```suggestion
```

File: src/adv_diff/BrinkmanAdvDiffSemiImplicitHierarchyIntegrator.cpp [hunk 14]
```diff
@@ -104,14 +104,6 @@ BrinkmanAdvDiffSemiImplicitHierarchyIntegrator::BrinkmanAdvDiffSemiImplicitHiera
     bool register_for_restart)
     : AdvDiffSemiImplicitHierarchyIntegrator(object_name, input_db, register_for_restart)
 {
-#if !defined(NDEBUG)
-    TBOX_ASSERT(!object_name.empty());
-    TBOX_ASSERT(input_db);
-#endif
```
Comment (drwells) [inline src/adv_diff/BrinkmanAdvDiffSemiImplicitHierarchyIntegrator.cpp:110]: Why was this removed?

**Review by drwells — CHANGES_REQUESTED** [review]
We're still missing the items on the checklist: i.e., there needs to be a changelog entry for the incompatibility and also new tests for the new (time-dependent) functionality.

File: src/adv_diff/BrinkmanAdvDiffSemiImplicitHierarchyIntegrator.cpp (hunk 14 above)
Comment (Raamkrishnan) [reply src/adv_diff/BrinkmanAdvDiffSemiImplicitHierarchyIntegrator.cpp:110]: Because these conditions are checked in the ```AdvDiffSemiImplicitHierarchyIntegrator``` class constructor. Here it was redundant.

Comment (boyceg) [reply src/adv_diff/BrinkmanAdvDiffSemiImplicitHierarchyIntegrator.cpp:110]: I think it is better to keep the assertions if they are relevant for the current method, even if they are checked elsewhere --- they shouldn't cause any harm here. Also, although it seems unlikely, they also could (accidentally or intentionally) be removed from the other implementation.

Comment (Raamkrishnan) [reply src/adv_diff/BrinkmanAdvDiffSemiImplicitHierarchyIntegrator.cpp:110]: Sounds good.

**Comment by Raamkrishnan** [comment]
I have added a test for the time-dependent advection-diffusion problem (```examples/adv_diff/ex8```).  Also changelog is added. Please take a look.

File: include/ibamr/BrinkmanAdvDiffSemiImplicitHierarchyIntegrator.h (hunk 1 above)
Comment (drwells) [reply include/ibamr/BrinkmanAdvDiffSemiImplicitHierarchyIntegrator.h]: I think we settled on `setTransportQuantityTimeIndependent()`: hence this is resolved.

**Review by drwells — APPROVED** [review]

**Comment by drwells** [comment]
@amneetb you still have some outstanding requests - are you happy with the current state of things?

**Review by amneetb — APPROVED** [review]
